lib/router/router.js

```diff
@@ -121,11 +121,13 @@ export default class Router extends EventEmitter {
   }
 
   push (url, as = url, options = {}) {
-    return this.change('pushState', url, as, options)
+    const formattedUrl = typeof url === 'object' ? format(url) : url
+    return this.change('pushState', formattedUrl, as === url ? formattedUrl : as, options)
```

Comment:
I think it's better to have something like `formattedAs` too. It makes code clear and simple.